consensus/src/creation/mod.rs

```diff
@@ -133,86 +108,97 @@ async fn keep_processing_units_until<H: Hasher>(
+            error!(target: LOG_TARGET, "Starting round not provided: {}", e);
             return;
         }
     };
 
-    if let Err(err) = run_creator(conf, io, starting_round).await {
+    if let Err(err) = run_creator(conf, io, keychain, starting_round).await {
         match err {
             CreatorError::OutChannelClosed(e) => {
-                warn!(target: "AlephBFT-creator", "Notification send error: {}. Exiting.", e)
+                warn!(target: LOG_TARGET, "Notification send error: {}. Exiting.", e)
             }
             CreatorError::ParentsChannelClosed => {
-                debug!(target: "AlephBFT-creator", "Incoming parent channel closed, exiting.")
+                debug!(target: LOG_TARGET, "Incoming parent channel closed, exiting.")
             }
         }
     }
 }
 
-async fn run_creator<H: Hasher>(
+async fn run_creator<H: Hasher, D: Data, MK: MultiKeychain, DP: DataProvider<D>>(
     conf: Config,
-    io: &mut IO<H>,
+    io: &mut IO<H, D, MK, DP>,
+    keychain: MK,
     starting_round: Round,
 ) -> anyhow::Result<(), CreatorError> {
-    let Config {
-        node_id,
-        n_members,
-        create_lag,
-        max_round,
-    } = conf;
+    let node_id = conf.node_ix();
+    let n_members = conf.n_members();
+    let create_delay = conf.delay_config().unit_creation_delay.clone();
+    let max_round = conf.max_round();
+    let session_id = conf.session_id();
     let mut creator = Creator::new(node_id, n_members);
+    let packer = Packer::new(keychain, session_id);
     let incoming_parents = &mut io.incoming_parents;
     let outgoing_units = &io.outgoing_units;
+    let data_provider = &mut io.data_provider;
 
-    debug!(target: "AlephBFT-creator", "Creator starting from round {}", starting_round);
+    debug!(target: LOG_TARGET, "Creator starting from round {}", starting_round);
     for round in starting_round..max_round {
         // Skip waiting if someone created a unit of a higher round.
         // In such a case at least 2/3 nodes created units from this round so we aren't skipping a
         // delay we should observe.
         let skip_delay = creator.current_round() > round;
         if !skip_delay {
-            let lag = Delay::new(create_lag(round.into()));
+            let delay = Delay::new(create_delay(round.into()));
 
-            keep_processing_units_until(&mut creator, incoming_parents, lag).await?;
+            keep_processing_units_until(&mut creator, incoming_parents, delay).await?;
         }
 
-        let (unit, parent_hashes) = create_unit(round, &mut creator, incoming_parents).await?;
-
-        trace!(target: "AlephBFT-creator", "Created a new unit {:?} at round {:?}.", unit, round);
+        let (preunit, parent_hashes) = create_unit(round, &mut creator, incoming_parents).await?;
+        trace!(target: LOG_TARGET, "Created a new preunit {:?} at round {:?}.", preunit, round);
+        let data = data_provider.get_data().await;
```

Comment:
Concerns:
1) this could be called parallel to pre-unit creation - these are independent tasks + calls to .get_data slows down units processing throughput of the `creator`
2) would be a good idea to have some common and simple pattern for calling `external` api implementations, which can potentially block indefinitely (this one even uses a Mutex...in a safe way :) )

Both could be solved by wrapping a provided impl of `DataProvider` by something that for example spawns_blocking and buffers calls to `.get_data`.

Reply:
Both of these are actually on purpose:
1. We want to have the freshest possible data, while `create_unit` is expected to take quite a while in a normal situation, it has an explicit delay. This had an explicit comment beforehand, should I re-add it?
2. We want this to be able to block, because then you can have a "lazy" consensus mechanism that only does any work when data is available (or not yet ordered, the implementation of such a thing is not completely trivial).